### DISCUSSION TITLE
Syntax Highlighting for Rust doc comment

### DISCUSSION THREAD

**Discussion by MunifTanjim** [nvim-treesitter/nvim-treesitter#3521] [Q&A]
Rust comments: https://doc.rust-lang.org/reference/comments.html

It looks like this in source code: https://github.com/matklad/once_cell/blob/a0aeb9b3780dde7f9523bb78755b3d70cd1d2657/src/lib.rs#L354-L377

```rs
    /// A cell which can be written to only once. It is not thread safe.
    ///
    /// Unlike [`std::cell::RefCell`], a `OnceCell` provides simple `&`
    /// references to the contents.
    ///
    /// [`std::cell::RefCell`]: https://doc.rust-lang.org/std/cell/struct.RefCell.html
    ///
    /// # Example
    /// ```
    /// use once_cell::unsync::OnceCell;
    ///
    /// let cell = OnceCell::new();
    /// assert!(cell.get().is_none());
    ///
    /// let value: &String = cell.get_or_init(|| {
    ///     "Hello, World!".to_string()
    /// });
    /// assert_eq!(value, "Hello, World!");
    /// assert!(cell.get().is_some());
    /// ```
    pub struct OnceCell<T> {
        // Invariant: written to at most once.
        inner: UnsafeCell<Option<T>>,
    }
```

It's basically a markdown content where each line is prefixed with `/// `.

Now the question is what needs to be done to be able to enable syntax highlighting for these?

Is this ever going to be supported (assuming it needs changes in `nvim-treesitter`, maybe even in neovim core...)?

---

Related Links:
- https://github.com/nvim-treesitter/nvim-treesitter/issues/2212
- https://github.com/neovim/neovim/pull/17099
- https://github.com/neovim/neovim/pull/19563

**Comment by clason** [comment]
You could try a Markdown injection with the `#trim` directive?

**Reply by MunifTanjim** [reply]
Hmm... I think `#trim` is not useful here?

https://github.com/nvim-treesitter/nvim-treesitter/blob/539bed2f4418f193618994b198c70f36212501f0/lua/nvim-treesitter/query_predicates.lua#L157-L159

Looks like it just removes empty lines.

---

So far, came up with this:

```scm
((((line_comment) @_line (#lua-match? @_line "^///")) @combined) @markdown)
```

Not really sure how `@combine` behaves...
- Does it combines only the comment lines that are together?
- Or does it combines all the lines matching lines in the file?

Proper query would be to use [Quantification Operators](https://tree-sitter.github.io/tree-sitter/using-parsers#quantification-operators), but those are not supported yet.

I've added some links I could find related to this discussion.

**Reply by clason** [reply]
Ah, might be "offset" (on mobile right now).

**Reply by theHamsta** [reply]
The problem that the implementation is a bit buggy regarding injection with gaps and also `@combined`. I  noticed another time when I tried to implement this using the `#exclude-children!`.

It would make a lot of stuff easier when we would drop all "non-combined" as it's also not supported by upstream tree-sitter.

The injection of markdown into rust comments is supposed to work without any directives when following the behavior of upstream tree-sitter:
 - successive markdown fragments would always be one document
 - /// gets automatically excluded because it is an (anonymous) child if the injection node

Even if we don't have this as the default behavior now it is supposed to work with `@combined`/`@exclude-children` (there should be an open draft request, trying to use it also for anonymous nodes). I'd think it would be easiest to modify the code in core when it's the upstream default behavior instead of trying to stack multiple complex implementation together and also quantificators. Alternatively, we should test our current directives to see why they don't yield the expected result yet. This problem affects álso all template languages.

When each parser always seems the document as whole we could just forward all edits to all parsers and would have incremental parsing and also a proper implementation `@combined` vs the one that we have now which seems be sometimes a bit off (because it is rarely used? Difficult to see when injections ranges are one char wrong?). @muniftanjim question on whether the parser sees the line end depends on exactly such implementation details. Scanners have the possibility to find out whether they entered a new injection range.

For the question of @MunifTanjim 
- `@combined` for nvim means that for multiple ranges only one syntax tree should be created (if I remember correctly, not 100% sure, also on mobile). Upstream tree-sitter knows no other behavior than this one
- `#combine!` for upstream tree-sitter means that children of the injection language are excluded (a combined document, languages don't overlap, this is the default). The alternative setting is `#include-children!`. Here an injection node would have children in the original language and in the same ranges also of the injection (Don't know if I got names correct)
Both usages have in common that they influence the call set_included_ranges of a parser

**Reply by theHamsta** [reply]
So ideally `(comment) @markdown` would just work.

**Reply by DougAnderson444** [reply]
Hi there! I am also looking for this solution.

One thing to add is the solution should resolve the [Link] within the project, right?

**Reply by theHamsta** [reply]
> One thing to add is the solution should resolve the [Link] within the project, right?

What do you mean by resolve? By default, tree-sitter performs only syntax highlighting. There is no resolving of links (e.g. for `gx` yet)

**Reply by DougAnderson444** [reply]
What I mean is: text in brackets will only be highlighted if the text leads to an actual file. 

So [AValidLink] would highlight as a link, whereas [DoesNotExist] would not be highlighted as a link.

This is what VS Code does. I'm not sure if it's already built into rust-analyzer, could be? I think missing links give compiler errors, so it's likely.

Thought I'd mention it as it would be great to achieve feature parity with vscode

**Reply by clason** [reply]
Apples and oranges. This has nothing to do with tree-sitter, which is a purely syntax parser. You're comparing a full editor with extensions with a single specialized subsystem.
